### PR TITLE
Binary search tree

### PR DESCRIPTION
**Pull request by masadl**

### CODE REVIEW

**Comment by keisuke-umezawa** [comment]
aaaa

**Comment by AkiraMisawa** [comment]
🍻

File: BinarySearchTree/BinarySearchTree.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        void delete_subtree(const Node* subtree_root);
+
+        void insert_to_subtree(const int val, Node* subtree_root);
+
+        bool find_subtree(const int val, const Node* subtree_root);
+
+        void print_subtree_elems(const Node* subtree_root);
+
+    public:
+        Node* _pnt_to_root;
```
Comment (IgnorantCoder) [inline BinarySearchTree/BinarySearchTree.h]: change name. _root.

Comment (IgnorantCoder) [inline BinarySearchTree/BinarySearchTree.h]: use std::unique_ptr.

File: BinarySearchTree/Node.cpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
+//
+// Created by Masataka Ushiku on 2017/08/15.
+//
+
+#include "Node.h"
+#include <algorithm>
+
+namespace lib {
+    Node::Node() = default;
```
Comment (IgnorantCoder) [inline BinarySearchTree/Node.cpp]: need not default constructor.

File: BinarySearchTree/BinarySearchTree.h [hunk 3]
```diff
@@ -0,0 +1,36 @@
+//
+// Created by Masataka Ushiku on 2017/08/15.
+//
+
+#include "Node.h"
+
+#pragma once
```
Comment (IgnorantCoder) [inline BinarySearchTree/BinarySearchTree.h]: #pragma once must be top!!!

File: BinarySearchTree/Node.cpp [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    Node::Node(int val)
+     :_left_child(nullptr),
+     _right_child(nullptr),
+     _val(val)
+    {}
+
+    Node::~Node()
+    {}
+
+    Node::Node(const Node& other) = default;
```
Comment (IgnorantCoder) [inline BinarySearchTree/Node.cpp]: Maybe cannot use default, have raw pointer.

File: BinarySearchTree/Node.h [hunk 5]
```diff
@@ -0,0 +1,58 @@
+//
+// Created by Masataka Ushiku on 2017/08/15.
+//
+
+#pragma once
+
+namespace lib {
+    class Node {
```
Comment (IgnorantCoder) [inline BinarySearchTree/Node.h]: use composite pattern.

File: BinarySearchTree/Node.h [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    public:
+        /**
+         * @brief default constructor
+         */
+        Node();
```
Comment (IgnorantCoder) [inline BinarySearchTree/Node.h]: need not default constructor

File: BinarySearchTree/Node.cpp [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+    Node& Node::operator= (Node other)
+    {
+        return *this = std::move(Node(other));
```
Comment (IgnorantCoder) [inline BinarySearchTree/Node.cpp]: Where is move assignment operator?

**Pull request merged**